src/pubsub-async-iterator.ts

```diff
@@ -0,0 +1,110 @@
+import { $$asyncIterator } from 'iterall';
+import {PubSubEngine} from './pubsub-engine';
+
+/**
+ * A class for digesting PubSubEngine events via the new AsyncIterator interface.
+ * This implementation is a generic version of the one located at
+ * @class
+ *
+ * @constructor
+ *
+ * @property pullQueue @type {Function[]}
+ * A queue of resolve functions waiting for an incoming event which has not yet arrived.
+ * This queue expands as next() calls are made without PubSubEngine events occurring in between.
+ *
+ * @property pushQueue @type {any[]}
+ * A queue of PubSubEngine events waiting for next() calls to be made.
+ * This queue expands as PubSubEngine events arrice without next() calls occurring in between.
+ *
+ * @property eventsArray @type {string[]}
+ * An array of PubSubEngine event names which this PubSubAsyncIterator should watch.
+ *
+ * @property allSubscribed @type {Promise<number[]>}
+ * A promise of a list of all subscription ids to the passed PubSubEngine.
+ *
+ * @property listening @type {boolean}
+ * Whether or not the PubSubAsynIterator is in listening mode (responding to incoming PubSubEngine events and next() calls).
+ * Listening begins as true and turns to false once the return method is called.
+ *
+ * @property pubsub @type {PubSubEngine}
+ * The PubSubEngine whose events will be observed.
+ */
+export class PubSubAsyncIterator<T> implements AsyncIterator<T> {
+
+  private pullQueue: Function[];
+  private pushQueue: T[];
+  private eventsArray: string[];
+  private allSubscribed: Promise<number[]>;
+  private listening: boolean;
+  private pubsub: PubSubEngine;
+
+  constructor(pubsub: PubSubEngine, eventNames: string | string[]) {
+    this.pubsub = pubsub;
+    this.pullQueue = [];
+    this.pushQueue = [];
+    this.listening = true;
+    this.eventsArray = typeof eventNames === 'string' ? [eventNames] : eventNames;
+    this.allSubscribed = this.subscribeAll();
+  }
+
+  public async next() {
+    await this.allSubscribed;
+    return this.listening ? this.pullValue() : this.return();
+  }
+
+  public async return() {
+    this.emptyQueue(await this.allSubscribed);
+    return { value: undefined, done: true };
+  }
+
+  public async throw(error) {
+    this.emptyQueue(await this.allSubscribed);
+    return Promise.reject(error);
+  }
+
+  public [$$asyncIterator]() {
+    return this;
+  }
+
+  private async pushValue(event: T) {
+    await this.allSubscribed;
+    if (this.pullQueue.length !== 0) {
+      this.pullQueue.shift()({ value: event, done: false });
+    } else {
+      this.pushQueue.push(event);
+    }
+  }
+
+  private pullValue() {
+    return new Promise((resolve => {
+      if (this.pushQueue.length !== 0) {
+        resolve({ value: this.pushQueue.shift(), done: false });
+      } else {
+        this.pullQueue.push(resolve);
+      }
+    }).bind(this));
```

Comment:
`bind` not needed with arrow function.

Reply:
@AlexLeung were you able to run the tests successfully?

Reply:
Can we fix this lint?